examples/pytorch/graphsage/graphsage.py

```diff
@@ -0,0 +1,295 @@
+"""
+Inductive Representation Learning on Large Graphs
+Paper: http://papers.nips.cc/paper/6703-inductive-representation-learning-on-large-graphs.pdf
+Code: https://github.com/williamleif/graphsage-simple
+Simple reference implementation of GraphSAGE.
+"""
+import argparse, time, math
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from dgl import DGLGraph
+from dgl.data import register_data_args, load_data
+
+
+def graphsage_msg(edge):
+    msg = edge.src['h']
+    return {'m': msg}
+
+
+class Aggregator(nn.Module):
+    def __init__(self, g, in_feats, out_feats, activation=None, bias=True):
+        super(Aggregator, self).__init__()
+        self.g = g
+        self.weight = nn.Parameter(torch.Tensor(in_feats, out_feats))  # (F,EF)
+        if bias:
+            self.bias = nn.Parameter(torch.Tensor(out_feats))  # (EF,1)
+        else:
+            self.bias = None
+        self.activation = activation
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        weight_stdv = 1. / math.sqrt(self.weight.size(1))
+        self.weight.data.uniform_(-weight_stdv, weight_stdv)
+        if self.bias is not None:
+            bias_stdv = 1. / math.sqrt(self.bias.size(0))
+            self.bias.data.uniform_(-bias_stdv, bias_stdv)
+
+    def forward(self, node):
+        nei = node.mailbox['m']  # (B, N, F)
+        h = node.data['h']  # (B, F)
+        h = self.concat(h, nei, node)  # (B, F)
+        h = torch.mm(h, self.weight)  # (B, EF)
+        if self.bias is not None:
+            h = h + self.bias
+        if self.activation:
+            h = self.activation(h)
+        return {'h': h}
+
+    def concat(self, h, nei, nodes):
+        print('no implementation !')
+        pass
+
+
+class MeanAggregator(Aggregator):
+    def __init__(self, g, in_feats, out_feats, activation, bias):
+        super(MeanAggregator, self).__init__(g, in_feats, out_feats, activation, bias)
+
+    def concat(self, h, nei, nodes):
+        degs = self.g.in_degrees(nodes.nodes()).float()
+        if h.is_cuda:
+            degs = degs.cuda(h.device)
+        concatenate = torch.cat((nei, h.unsqueeze(1)), 1)
+        concatenate = torch.sum(concatenate, 1) / degs.unsqueeze(1)
+        return concatenate
+
+
+class PoolingAggregator(Aggregator):
+    def __init__(self, g, in_feats, out_feats, activation, bias):  # (2F, F)
+        super(PoolingAggregator, self).__init__(g, in_feats*2, out_feats, activation, bias)
+        self.mlp = PoolingAggregator.MLP(in_feats, in_feats, F.relu, False, True)
+
+    def concat(self, h, nei, nodes):
+        nei = self.mlp(nei)
+        concatenate = torch.cat((nei, h), 1)
+        return concatenate
+
+    class MLP(nn.Module):
```

Comment:
Move this outside the `PoolingAggregator` class.

Reply:
Thanks for your review!! I learned a lot!
 I wonder why it is better?

Reply:
Nested/Inner class should usually be avoided. To me, there is no benefit. In terms of encapsulation, I could still access the inner class as a class attribute like `PoolingAggregator.MLP`. So why not just `MLP`? Usually when I saw a nested class I will try to think "hmm... why is it needed? is there any magic?", but the answer is usually no.